<a href="https://colab.research.google.com/github/NatFT/PythonCourse/blob/main/Numpy_mini_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Investigate, wrangle, add new data to the dataframe, filter and correlate

The following data file contains data about student scores in math, reading and writing.  The exam scores are assumed to be percentages. It also contains data about gender, ethnicity, parental education, whether the student qualifies for free school food and whether or not the student has taken a preparation course for the exams.  The data set has already been cleaned. 

1.  Investigate the data set.

2.  Create numpy arrays to hold each of the three sets of scores.  
Create a new numpy array to hold the average exam score (of the three scores for each student)
Add the new numpy array as a new column in the dataframe.

3.  Filter the original dataset into a new dataframe containing just the females.  Calculate the average exam scores for all rows in this new dataframe and then find the mean of the average column.

4.  Do the same for the males.

5.  Use the original dataset to find the correlation coefficient for reading and maths.  How closely do they correlate?  Write what you find in a text box below the code.

6.  Do the same for reading and writing. What do you find?

7.  You might want to filter on different criteria and check correlation (e.g. those on free school meals, or those who had prepared, etc)







The dataset can be accessed here:  https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv.  This is a .csv file.

**NOTE:** Some useful references are included at the bottom of this spreadsheet.

Use the code cell below to work your code.

In [30]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv"
df = pd.read_csv(url)
# df.info()
# display(df)

#create arrays for each score
maths_arr = np.array(df['math_score'], np.int64)
reading_arr = np.array(df['reading_score'], np.int64)
writing_arr = np.array(df['writing_score'], np.int64)

#claculate average exam score for each student
avg_exam_score = (maths_arr + reading_arr + writing_arr) / 3
df['average_exam_score'] = avg_exam_score.tolist()
# display(df)

#calculate the overall mean of the average exam scores for the female students
df_female = df[df['gender'] == 'female']
# display(df_female)
# df_female.info()
avg_score_f = np.array(df_female['average_exam_score'], np.float64)
avg_score_f_mean = avg_score_f.mean()
print(avg_score_f_mean)

#calculate the overall mean of the average exam scores for the male students
df_male = df[df['gender'] == 'male']
# display(df_male)
# df_male.info()
avg_score_m = np.array(df_male['average_exam_score'], np.float64)
avg_score_m_mean = avg_score_m.mean()
# print(avg_score_m_mean)

#the correlation between maths and reading exam scores
coef = np.corrcoef(maths_arr, reading_arr)
# print(coef)

#the correlation between reading and writing exam scores
coef = np.corrcoef(writing_arr, reading_arr)
print(coef)

df_free_lunch = df[df['lunch'] == 'free/reduced']
df_standard_lunch = df[df['lunch'] == 'standard']

avg_score_free_lunch = np.array(df_free_lunch['average_exam_score'], np.float64)
avg_score_standard_lunch = np.array(df_standard_lunch['average_exam_score'], np.float64)

free_min = avg_score_free_lunch.min()
free_mean = avg_score_free_lunch.mean()
free_max = avg_score_free_lunch.max()
print(free_min, free_mean, free_max)

standard_min = avg_score_standard_lunch.min()
standard_mean = avg_score_standard_lunch.mean()
standard_max = avg_score_standard_lunch.max()
print(standard_min, standard_mean, standard_max)

diff_mean = standard_mean - free_mean
print(diff_mean)

70.81989247311829
[[1.         0.95676863]
 [0.95676863 1.        ]]
13.666666666666666 62.3399433427762 95.33333333333333
17.333333333333332 71.68624420401855 100.0
9.346300861242348


The students maths and reading scores have a correlation coefficient of 0.81 (to 2dp). Therefore, there is a strong positive correlation, i.e. if a student does well in their maths exam they are likely to do well in their reading exam.

The students reading and writing scores have a correlation coefficient of 0.96 (to 2dp). Therefore, there is a very strong positive correlation, i.e. if a student does well in their reading exam they are very likely to do well in their writing exam.

### Helpful references
---

Filtering on criteria using df references:  
`filtered_df = df[df['column name' == value]]`  

Filtering on multiple criteria using df reference:  
`filtered_df = df[df['first column name' == value] & df[second column name] >= value]`

Numpy help sheet:  http://datacamp-community-prod.s3.amazonaws.com/da466534-51fe-4c6d-b0cb-154f4782eb54

